### PR TITLE
calibre sax json parser

### LINKED ISSUE #9016: Calibre Metadata Scan only finds 1 book (Kobo Clara HD)
* KOReader version: 
* Device: Kobo Clara HD

#### Issue
Have completely switched over to koreader from kobo/nickel and am enjoying the software very much.

One issue is with the search calibre metadata functionality (i.e. browse series), despite deleting and having calibre recreate the driveinfo.calibre and metadata.calibre files.  The metadata in calibre exists and is in good order.

When I use the calibre->search settings -> manage libraries -> rescan disk for calibre libraries option, the resulting message says, "Found 1 calibre library with 1 books:  1: /mnt/onboard 

The actual calibre library (and the device) holds 3261 books.

##### `crash.log` (if applicable)

The following crash log is of an actual crash, but repeating the attempt to scan the calibre libraries does not always generate the crash.

<details>
  <summary>crash.log</summary>

```
04/17/22-14:59:12 WARN  cache is older than metadata, ignoring it
04/17/22-14:59:38 INFO  calibre info loaded from disk (search) in 26244.237 milliseconds: 1 books
04/17/22-14:59:39 INFO  metadata: 1 books imported from calibre in 28796.470 milliseconds
./luajit: not enough memory
!!!!
Uh oh, something went awry... (Crash n°1: 04/17/22 @ 14:59:43)
Running FW 4.30.18838 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
Attempting to restart KOReader . . .
!!!!
```
</details>

Thanks for any insight into this issue.

### LINKED ISSUE #11215: Calibre metadata search: crash
* KOReader version: 2023.10
* Device: Pocketbook Era

#### Issue
Clicking on an entry in the search results list leads to a crash.

#### Steps to reproduce
Calibre metadata search -> Enter search text -> Click on an entry in the search results list

[crash.log](https://github.com/koreader/koreader/files/13627476/crash.log)

### LINKED ISSUE #11611: No tags under "calibre metadata search"
* KOReader version: v2024.03.01
* Device: Kobo Libra 2

#### Issue
When I go to _Calibre metadata search_ then select _Browse tags_, there are no tags populating.

I have tried reinstalling koreader from scratch (just copying over my dictionaries to the new install), and the issue persists.

My home folder for koreader is /mnt/onboard/Books.  I use this same folder setup for my mother's Kobo, and hers will populate tags.  Her Calibre library is a smaller subset of mine, but lives on the same MacBook.

I have no idea how to begin troubleshooting this, other than the fresh install of koreader that I already tried.

Any help is greatly appreciated!

### PR DESCRIPTION
**Pull request by pazos**
Add a proper sax parser for "calibre.metadata" files, using lunajson.

Requires https://github.com/koreader/koreader-base/pull/1801

Fixes #11611
Fixes #11215
Fixes #9016

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11922)

### CODE REVIEW

File: plugins/calibre.koplugin/parser.lua [hunk 1]
```diff
@@ -1,90 +1,137 @@
...
+local function append(s)
+   if isField(s) then
+      wanted = false
+      if isRequiredField(s) then
+         wanted = true
+      end
+   else
+      if wanted_struct then
+         table.insert(t[last_key], s)
+      elseif wanted and s and last_key then
```
Comment (pazos) [inline plugins/calibre.koplugin/parser.lua:92]: fix me

File: plugins/calibre.koplugin/metadata.lua [hunk 2]
```diff
@@ -54,7 +54,7 @@ end
 
 -- this is the max file size we attempt to decode using json. For larger
 -- files we want to attempt to manually parse the file to avoid OOM errors
-local MAX_JSON_FILESIZE = 30 * 1000 * 1000
+local MAX_JSON_FILESIZE = 30
```
Comment (pazos) [inline plugins/calibre.koplugin/metadata.lua:57]: fix me (and probably update to 50MB?)

Comment (Frenzie) [reply plugins/calibre.koplugin/metadata.lua:57]: Maybe more like this?
https://github.com/koreader/koreader/blob/c7a59145a307b1de0e3843ffc00c541c8746999f/defaults.lua#L14-L21
https://github.com/koreader/koreader/blob/c7a59145a307b1de0e3843ffc00c541c8746999f/frontend/document/doccache.lua#L15-L21

**Comment by pazos** [comment]
@Frenzie: what is the proper magic to exclude lunajson files from CI?

File: plugins/calibre.koplugin/lunajson/encoder.lua [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+local setmetatable = setmetatable
+local huge, tiny = 1/0, -1/0
+
+local f_string_esc_pat
+if _VERSION == "Lua 5.1" then
+	-- use the cluttered pattern because lua 5.1 does not handle \0 in a pattern correctly
+	f_string_esc_pat = '[^ -!#-[%]^-\255]'
+else
+	f_string_esc_pat = '[\0-\31"\\]'
+end
```
Comment (NiLuJe) [inline plugins/calibre.koplugin/lunajson/encoder.lua:15]: Random drive-by: LuaJIT should support something like `[%z\1-\31"\\]'

File: plugins/calibre.koplugin/lunajson/decoder.lua [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+	error("parse error at " .. pos .. ": " .. errmsg, 2)
+end
+
+local f_str_ctrl_pat
+if _VERSION == "Lua 5.1" then
+	-- use the cluttered pattern because lua 5.1 does not handle \0 in a pattern correctly
+	f_str_ctrl_pat = '[^\32-\255]'
+else
+	f_str_ctrl_pat = '[\0-\31]'
+end
```
Comment (NiLuJe) [inline plugins/calibre.koplugin/lunajson/decoder.lua:20]: Ditto, `[%z\1-\31]` ought to work ;).

Comment (NiLuJe) [reply plugins/calibre.koplugin/lunajson/decoder.lua:20]: (c.f., `ffi/sha2.lua` in base for a fancy is_LuaJIT check ;).)

**Comment by pazos** [comment]
@NiLuJe: thanks for the review but I didn't expect one on foreign code :)

**Comment by Frenzie** [comment]
@pazos 
```diff
diff --git a/kodev b/kodev
index eb2e52d0f..cec1342de 100755
--- a/kodev
+++ b/kodev
@@ -962,7 +962,7 @@ function kodev-check() {
         exit_code=1
     fi
 
-    tab_detected=$(grep -P "\\t" --include \*.lua --exclude={dateparser.lua,xml.lua} --recursive {reader,setupkoenv,datastorage}.lua frontend plugins spec || true)
+    tab_detected=$(grep -P "\\t" --include \*.lua --exclude={dateparser.lua,lunajson.lua,xml.lua} --exclude-dir=lunajson --recursive {reader,setupkoenv,datastorage}.lua frontend plugins spec || true)
     if [ "${tab_detected}" ]; then
         echo -e "\\n${ANSI_RED}Warning: tab character detected. Please use spaces."
         echo "${tab_detected}"
```

**Comment by pazos** [comment]
The implementation was broken, see https://www.mobileread.com/forums/showthread.php?p=4427793#post4427793

Now it is the proper thing, AFAICT.

File: plugins/calibre.koplugin/metadata.lua (hunk 2 above)
Comment (pazos) [reply plugins/calibre.koplugin/metadata.lua:57]: I don't know. I think I would like to keep a fixed size. Most of the bug reports I get are from mobileread users and seems easier for them to tell me the size of their `metadata.calibre` files than to attach logs.

But, more important than that:
- current 30MB size is too big for rapidsjon to handle in some devices/configs, for instance https://github.com/koreader/koreader/issues/11215 ***
- a sax parser, even in plain lua, should be faster than a dom parser on a scenario where we care about less than a 1-10% of the file. ****

***: most likely a setup with few books and a huge amount of user metadata per book.

****: this holds true on my tests with a little library of ~400 books and no fancy/extra metadata per book. Even after I start a wireless connection, which strips all the unused metadata from the json file, I usually got reads in the same order of magnitude using lunajson sax parser vs rapidjson dom parser.

**Comment by Frenzie** [comment]
Alright, sounds good.

**Comment by pazos** [comment]
If some of you (wink, wink @NiLuJe) have a big on-device metadata with a bunch of custom columns I wouldn't mind a benchmark (or at least the nº of books / size in bytes of the document)

FWIW, here 150 *naked* books got me a `metadata.calibre` of ~620KiB, so less than 5KiB per book.

I'm assuming it could grow to +50 KiB per book. (On mobileread an advanced user said ~4000 books weight 40MiB, so maybe 50 is the worst case scenario)

Even then a `MAX_JSON_FILESIZE` of `10MiB` could handle 2000 books without extra metadata or 200 books with too much metadata.

So it seems a nice threshold to switch to the sax parser.

**Comment by benoit-pierre** [comment]
Is there an advantage in having 2 code paths? Would always using the sax parser make a noticeable (negative) impact on performance?

**Comment by pazos** [comment]
> Is there an advantage in having 2 code paths?

Very good question. That's why I would like a benchmark from somebody that uses calibre heavily and doesn't use wireless transfers to push their books.

> Would always using the sax parser make a noticeable (negative) impact on performance?

Not on a desktop with a few hundred books. Rapidjson is faster with all unneeded fields stripped and **not** slower with default calibre fields.

No idea about real devices since my metadata there is already stripped.

Anyhow, on my **quite limited** tests both produce results in the same order of magnitude, but rapidjson has less outliers.

**Comment by NiLuJe** [comment]
I don't really have fancy columns setup, but I'll try to give you some numbers ;).

**Comment by NiLuJe** [comment]
Yup, it's *considerably* slower on even a fairly modest library (metadata file is 7.2MB), even on a relatively speedy SoC (that was a Sage, it boosts its A7 cores fairly high, so much so that they might still be faster than the MTK's A53).

Anyway:

RapidJSON:

```
06/01/24-20:18:31 INFO  calibre info loaded from disk (search) in 472.000 milliseconds: 876 books 
06/01/24-20:18:31 INFO  metadata: 876 books imported from calibre in 499.000 milliseconds 
06/01/24-20:18:31 INFO  search done in 60.000 milliseconds (series, case sensitive: false, title: true, authors: true, series: false)
```

LunaJSON:

```
06/01/24-20:22:22 INFO  calibre info loaded from disk (search) in 1794.000 milliseconds: 876 books 
06/01/24-20:22:22 INFO  metadata: 876 books imported from calibre in 1820.000 milliseconds 
06/01/24-20:22:22 INFO  search done in 39.000 milliseconds (series, case sensitive: false, title: true, authors: true, series: false)
```

**Comment by NiLuJe** [comment]
On the upside, at a very quick glance, the results look correct ;).

**Comment by pazos** [comment]
> Yup, it's considerably slower even on a fairly modest library

Ok, thanks. That's much worse than I expected. And it doesn't seem like it is going to improve on very big libraries.

I see a few things that can be improved in this PR, I'm going to look.

**Comment by pazos** [comment]
# rapidjson (load + decode)
```
Parsed in 0.0059 milliseconds
Parsed in 0.0053 milliseconds
Parsed in 0.0048 milliseconds
Parsed in 0.0054 milliseconds
Parsed in 0.0045 milliseconds
Parsed in 0.0023 milliseconds
Parsed in 0.0023 milliseconds
Parsed in 0.0022 milliseconds
Parsed in 0.0022 milliseconds
Parsed in 0.0022 milliseconds
Parsed in 0.0022 milliseconds
Parsed in 0.0021 milliseconds
Parsed in 0.0021 milliseconds
Parsed in 0.0031 milliseconds
Parsed in 0.0036 milliseconds
Parsed in 0.0047 milliseconds
Parsed in 0.0051 milliseconds
Parsed in 0.0063 milliseconds
Parsed in 0.0051 milliseconds
Parsed in 0.0051 milliseconds
Parsed in 0.0045 milliseconds
Parsed in 0.0041 milliseconds
Parsed in 0.0038 milliseconds
Parsed in 0.0037 milliseconds
Parsed in 0.0034 milliseconds
Parsed in 0.0033 milliseconds
Parsed in 0.0031 milliseconds
Parsed in 0.0032 milliseconds
Parsed in 0.0041 milliseconds
Parsed in 0.0048 milliseconds
Parsed in 0.0041 milliseconds
Parsed in 0.0042 milliseconds
Parsed in 0.0056 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0049 milliseconds
Parsed in 0.0054 milliseconds
Parsed in 0.0053 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0051 milliseconds
Parsed in 0.0068 milliseconds
Parsed in 0.0060 milliseconds
Parsed in 0.0070 milliseconds
Parsed in 0.0074 milliseconds
Parsed in 0.0075 milliseconds
Parsed in 0.0062 milliseconds
Parsed in 0.0056 milliseconds
Parsed in 0.0049 milliseconds
Parsed in 0.0057 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0049 milliseconds
Parsed in 0.0079 milliseconds
Parsed in 0.0067 milliseconds
Parsed in 0.0058 milliseconds
Parsed in 0.0067 milliseconds
Parsed in 0.0091 milliseconds
Parsed in 0.0077 milliseconds
Parsed in 0.0082 milliseconds
Parsed in 0.0093 milliseconds
Parsed in 0.0078 milliseconds
Parsed in 0.0086 milliseconds
```

# lunajson

```
Parsed in 0.0104 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0101 milliseconds
Parsed in 0.0105 milliseconds
Parsed in 0.0106 milliseconds
Parsed in 0.0102 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0106 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0105 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0101 milliseconds
Parsed in 0.0122 milliseconds
Parsed in 0.0123 milliseconds
Parsed in 0.0109 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0101 milliseconds
Parsed in 0.0102 milliseconds
Parsed in 0.0121 milliseconds
Parsed in 0.0120 milliseconds
Parsed in 0.0190 milliseconds
Parsed in 0.0118 milliseconds
Parsed in 0.0107 milliseconds
Parsed in 0.0106 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0120 milliseconds
Parsed in 0.0124 milliseconds
Parsed in 0.0110 milliseconds
Parsed in 0.0105 milliseconds
Parsed in 0.0109 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0105 milliseconds
Parsed in 0.0110 milliseconds
Parsed in 0.0118 milliseconds
Parsed in 0.0126 milliseconds
Parsed in 0.0108 milliseconds
Parsed in 0.0119 milliseconds
Parsed in 0.0120 milliseconds
Parsed in 0.0105 milliseconds
Parsed in 0.0178 milliseconds
Parsed in 0.0123 milliseconds
Parsed in 0.0110 milliseconds
Parsed in 0.0124 milliseconds
Parsed in 0.0124 milliseconds
Parsed in 0.0110 milliseconds
Parsed in 0.0108 milliseconds
Parsed in 0.0107 milliseconds
Parsed in 0.0108 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0103 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0106 milliseconds
Parsed in 0.0125 milliseconds
Parsed in 0.0177 milliseconds
Parsed in 0.0187 milliseconds
Parsed in 0.0117 milliseconds
Parsed in 0.0124 milliseconds
Parsed in 0.0123 milliseconds
Parsed in 0.0118 milliseconds
Parsed in 0.0110 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0105 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0109 milliseconds
Parsed in 0.0110 milliseconds
Parsed in 0.0102 milliseconds
Parsed in 0.0123 milliseconds
Parsed in 0.0169 milliseconds
Parsed in 0.0117 milliseconds
Parsed in 0.0109 milliseconds
Parsed in 0.0106 milliseconds
Parsed in 0.0104 milliseconds
Parsed in 0.0105 milliseconds
```

**Comment by NiLuJe** [comment]
> I see a few things that can be improved in this PR, I'm going to look.

Give me a shout when you want me to re-test ;).

**Comment by pazos** [comment]
I'm not able to improve the thing too much, but at least catched a bug that would crash the program when trying to re-parse :)

Also improved a bit the tests, so they measure what is expected to measure:

# All calibre fields

```
# lunajson
Parsed in 0.0116 milliseconds
Parsed in 0.0101 milliseconds
Parsed in 0.0099 milliseconds
Parsed in 0.0094 milliseconds
Parsed in 0.0089 milliseconds
Parsed in 0.0092 milliseconds
Parsed in 0.0089 milliseconds
Parsed in 0.0093 milliseconds
Parsed in 0.0089 milliseconds
Parsed in 0.0091 milliseconds

# rapidjson
Parsed in 0.0015 milliseconds
Parsed in 0.0016 milliseconds
Parsed in 0.0015 milliseconds
Parsed in 0.0014 milliseconds
Parsed in 0.0015 milliseconds
Parsed in 0.0015 milliseconds
Parsed in 0.0015 milliseconds
Parsed in 0.0018 milliseconds
Parsed in 0.0014 milliseconds
Parsed in 0.0014 milliseconds
```

# Removed fields

```
# lunajson
Parsed in 0.0057 milliseconds
Parsed in 0.0052 milliseconds
Parsed in 0.0051 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0047 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0045 milliseconds
Parsed in 0.0046 milliseconds
Parsed in 0.0046 milliseconds

# rapidjson
Parsed in 0.0009 milliseconds
Parsed in 0.0010 milliseconds
Parsed in 0.0010 milliseconds
Parsed in 0.0009 milliseconds
Parsed in 0.0009 milliseconds
Parsed in 0.0008 milliseconds
Parsed in 0.0009 milliseconds
Parsed in 0.0008 milliseconds
Parsed in 0.0008 milliseconds
Parsed in 0.0009 milliseconds
```

**Comment by NiLuJe** [comment]
FWIW, something based on hashmap hitchecks instead of `ipairs` seems ever-so-slightly faster over here, and probably reads a tiny bit better?

```lua
-- parse "metadata.calibre" files
local lj = require("lunajson")

local array_fields = {
    authors = true,
    tags = true,
    series = true,
}

local required_fields = {
    authors = true,
    last_modified = true,
    lpath = true,
    series = true,
    series_index = true,
    size = true,
    tags = true,
    title = true,
    uuid = true,
}

local field
local t = {}
local function append(s)
    if array_fields[field] and t[field] then
        table.insert(t[field], s)
    else
        t[field] = s
    end
    field = nil
end

local depth = 0
local result = {}
local sax = {
    startobject = function()
        depth = depth + 1
    end,
    endobject = function()
        if depth == 1 then
            table.insert(result, t)
            t = {}
        end
        depth = depth - 1
    end,
    startarray = function()
        if array_fields[field] then
            t[field] = {}
        end
    end,
    key = function(s)
        if required_fields[s] then
            field = s
        end
    end,
    string = function(s)
        if not field then
            return
        end
        append(s)
    end,
    number = function(n)
        if not field then
            return
        end
        append(n)
    end,
    boolean = function(b)
        if not field then
            return
        end
        append(b)
    end,
    null = function()
        if not field then
            return
        end
        append()
    end,
}

local parser = {}
function parser.parseFile(file)
    result = {}
    local p = lj.newfileparser(file, sax)
    p.run()
    field = nil
    return result
end

return parser
```

(hashmap)
```
06/02/24-07:59:30 INFO  calibre info loaded from disk (search) in 1661.000 milliseconds: 876 books 
06/02/24-07:59:30 INFO  metadata: 876 books imported from calibre in 1686.000 milliseconds 
06/02/24-07:59:30 INFO  search done in 40.000 milliseconds (series, case sensitive: false, title: true, authors: true, series: false) 
```

vs.

(arrays + ipairs checks)
```
06/02/24-07:56:05 INFO  calibre info loaded from disk (search) in 1710.000 milliseconds: 876 books 
06/02/24-07:56:05 INFO  metadata: 876 books imported from calibre in 1736.000 milliseconds 
06/02/24-07:56:05 INFO  search done in 39.000 milliseconds (series, case sensitive: false, title: true, authors: true, series: false)
```

**Comment by NiLuJe** [comment]
Eeeeeh, most of this might simply come from not calling `append` at all for unwanted keys...

**Comment by NiLuJe** [comment]
Fixed only using the first value of an array in the above; I broke it at some point of the edits ;p.

**Comment by pazos** [comment]
Indeed, it looks much better :)

> Eeeeeh, most of this might simply come from not calling append at all for unwanted keys...

Yup, but that's the best we can get :)

> Is there an advantage in having 2 code paths? Would always using the sax parser make a noticeable (negative) impact on performance?

Based on the tests, yeah, we want two code paths. We also want to enforce `rapidjson` whenever it is possible. We could add an UI option to switch between implementations (keeping auto as default), so people could change to lunajson if they get OOM error with default settings or force `rapidjson` beyond the threshold if they know they device has enough ram to handle their particular json files.

**Comment by pazos** [comment]
> We could add an UI option to switch between implementations

For example:
Sin nombre

File: plugins/calibre.koplugin/parser.lua [hunk 5]
```diff
@@ -1,90 +1,93 @@
...
+}
 
-local function getValue(str, key)
-    if str == string.format("    \"%s\": null, ", key) then
-        return nil
+local field
+local t = {}
+local function append(v)
+    -- These *may* be arrays, so we need the extra check to confirm that startarray ran
+    if array_fields[field] and t[field] then
```
Comment (NiLuJe) [inline plugins/calibre.koplugin/parser.lua:26]: I realized we can simplify that some more ;).

```suggestion
    -- Some fields *may* be arrays, so check whether we ran through startarray first or not
    if t[field] then
```

File: plugins/calibre.koplugin/parser.lua [hunk 6]
```diff
@@ -1,90 +1,93 @@
...
+    boolean = function(b)
+        if field then
+            append(b)
+        end
+    end,
+    null = function()
+        if field then
+            append(nil)
+        end
+    end,
```
Comment (NiLuJe) [inline plugins/calibre.koplugin/parser.lua:81]: It also doesn't make much of a difference, but I figure we can also get rid of the `null` handler, as `t[field] = nil` is effectively a NOP.

Comment (NiLuJe) [reply plugins/calibre.koplugin/parser.lua:81]: (And LunaJSON will use an actual nop if a handler isn't set in the sax table).

File: plugins/calibre.koplugin/main.lua [hunk 7]
```diff
@@ -414,4 +421,52 @@ function Calibre:getWirelessMenuTable()
...
+                return G_reader_settings:hasNot("calibre_json_parser")
+            end,
+            callback = function()
+                G_reader_settings:delSetting("calibre_json_parser")
+            end,
+            hold_callback = function()
+                UIManager:show(InfoMessage:new{
+                    text = _("The program will decide based on the size of the JSON file. Recommended"),
+                })
+            end,
```
Comment (NiLuJe) [inline plugins/calibre.koplugin/main.lua:438]: Err, couldn't those simply be a `help_text`? instead of reimplementing part of that manually? ^^.

Comment (pazos) [reply plugins/calibre.koplugin/main.lua:438]: Lol, I didn't know that.

Comment (NiLuJe) [reply plugins/calibre.koplugin/main.lua:438]: I won't be the one blaming you for not being at ease with our UI widgets, it's really not my area of expertise either ;).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 5 of 7 files at r3, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11922)* status: all files reviewed, 3 unresolved discussions (waiting on @pazos)

___
*[`plugins/calibre.koplugin/metadata.lua` line 136 at r5](https://reviewable.io/reviews/koreader/koreader/11922#-NzTuh1Y2mytNQwq-zSh:-NzTuh1Y2mytNQwq-zSi:b-q37hlu) ([raw file](https://github.com/koreader/koreader/blob/230ee7d92b5c3d8b1317fa06931aa68a7321a66a/plugins/calibre.koplugin/metadata.lua#L136)):*
> ```Lua
>         else
>             books, err = rapidjson.load(self.metadata)
>         end
> ```

DRYer:

```
local impl = G_reader_settings:readSetting("calibre_json_parser") or attr.size > MAX_JSON_FILESIZE and "sax" or "dom"
if impl == "dom" then
    books, err = rapidjson.load(self.metadata)
else
    books, err = parser.parseFile(self.metadata)
end
```



_Code quote:_
```Lua
    if G_reader_settings:has("calibre_json_parser") then
        local impl = G_reader_settings:readSetting("calibre_json_parser")
        if impl == "dom" then
            books, err = rapidjson.load(self.metadata)
        elseif impl == "sax" then
            books, err = parser.parseFile(self.metadata)
        end
    else
        if attr.size > MAX_JSON_FILESIZE then
            books, err = parser.parseFile(self.metadata)
        else
            books, err = rapidjson.load(self.metadata)
        end
```

**Review by pazos — COMMENTED** [review]
I have no idea how to use this tool. Hopefully everything is ok :)

> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11922#-:-NzZeHdn-YV2cERcTf06:b-lcakq4)* status: 6 of 8 files reviewed, 3 unresolved discussions (waiting on @NiLuJe)

___
*[`plugins/calibre.koplugin/metadata.lua` line 136 at r5](https://reviewable.io/reviews/koreader/koreader/11922#-NzTuh1Y2mytNQwq-zSh:-NzZe85qAP3ZXWMyuFmq:b-896fix) ([raw file](https://github.com/koreader/koreader/blob/230ee7d92b5c3d8b1317fa06931aa68a7321a66a/plugins/calibre.koplugin/metadata.lua#L136)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

DRYer:

```
local impl = G_reader_settings:readSetting("calibre_json_parser") or attr.size > MAX_JSON_FILESIZE and "sax" or "dom"
    if impl == "dom" then
        books, err = rapidjson.load(self.metadata)
    else
        books, err = parser.parseFile(self.metadata)
end
```
</blockquote></details>

Done.
___
*[`plugins/calibre.koplugin/lunajson/decoder.lua` line 20 at r2](https://reviewable.io/reviews/koreader/koreader/11922#gh-1619324046:-NzZeBWL7QPVpBh8h76D:b-896fix) ([raw file](https://github.com/koreader/koreader/blob/f1a75844d1665f1301f5ae828a9f41f438de48af/plugins/calibre.koplugin/lunajson/decoder.lua#L20)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

(c.f., `ffi/sha2.lua` in base for a fancy is_LuaJIT check ;).)
</blockquote></details>

Done.
___
*[`plugins/calibre.koplugin/lunajson/encoder.lua` line 15 at r2](https://reviewable.io/reviews/koreader/koreader/11922#gh-1619322332:-NzZeCYNAB6QDOUO1RY9:b-896fix) ([raw file](https://github.com/koreader/koreader/blob/f1a75844d1665f1301f5ae828a9f41f438de48af/plugins/calibre.koplugin/lunajson/encoder.lua#L15)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

Random drive-by: LuaJIT should support something like `[%z\1-\31"\\]`
</blockquote></details>

Done.

**Review by NiLuJe — APPROVED** [review]
LGTM, thanks ;).

> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11922#-:-Nz_POP886BRt9KZ2UoK:b-4curuy)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @pazos)

**Comment by NiLuJe** [comment]
FWIW, I looked into what was happening with lua-rapidjson, because RapidJSON itself *has* a SAX API.

I remembered that it wasn't actually exposed as such because of performance reasons (calling Lua callbacks from C is *very* expensive, and those callbacks are called *a lot*), so I went ahead and poked at it to try to figure out where it could be using RAM...

Turns out... I... dunno, really: lua-rapidjson is actually using the SAX API internally, *and* it's reading the file in chunks already.

Sooo, it actually looks like it's actually the sheer amount of k,v pairs that just balloons to stupid amounts, so there wasn't much I could do. (I mean, I updated RapidJSON, enabled NEON, and switched to more sensible chunk sizes, and I cadged like 40ms this way, but I digress ;p).

Then, I kinda went: fuck it, let's try to do the "parsing" in C++ directly:

```
06/05/24-01:04:48 INFO  calibre info loaded from disk (search) in 268.000 milliseconds: 876 books 
06/05/24-01:04:48 INFO  metadata: 876 books imported from calibre in 295.000 milliseconds 
```

It *seems* to consume less memory, but, then again, as I mentioned earlier, the current approach really doesn't seem to bad on my library...

I'll let that sit for a few days and PR it later ;).

**Comment by NiLuJe** [comment]
Code lives in https://github.com/NiLuJe/lua-rapidjson if you're really curious.

**Comment by pazos** [comment]
>  if you're really curious.

I was :)

> Then, I kinda went: fuck it, let's try to do the "parsing" in C++ directly:

> 06/05/24-01:04:48 INFO  calibre info loaded from disk (search) in 268.000 milliseconds

That's impresive :))

> I'll let that sit for a few days and PR it later ;).

Please do :)
This PR is more than happy to be superseeded :). Also less UI, less noise.

Do you plan to upstream changes or keep a fork?

**Comment by NiLuJe** [comment]
> This PR is more than happy to be superseeded :). Also less UI, less noise.

I'd rather this make it in, actually. It might very well still be more memory-efficient, so it's still nice to have ;).

> Do you plan to upstream changes or keep a fork?

I'll PR the upstreamable bits, but we'll ultimately live with a fork because the calibre bits obviously don't make sense upstream.

FWIW, we're not losing much, the last upstream commit is already one of ours ;).

**Comment by NiLuJe** [comment]
And while we're in a data collecting mood: skipping the `slim`ming loops from `CalibreMetadata:cleanUnused` doesn't appear to make much of a difference, so that's good to know ;).

**Comment by NiLuJe** [comment]
> it actually looks like it's actually the sheer amount of k,v pairs that just balloons to stupid amounts

Actually... large string values *could* explain some of the overhead (e.g., the most common one being `description`).

**Comment by NiLuJe** [comment]
> I'd rather this make it in, actually.

Although, perhaps with slighty different menu labels/config keys. I'm thinking something along the lines of:

* Fast (RapidJSON via load_calibre)
* Safe (LunaJSON)
* Legacy (RapidJSON via load)

(And, for now, until we have more data from insane libraries, keep the size check, but perhaps defaultto fast in order to exercize the codepath and fund potential issues with it).

**Comment by pazos** [comment]
> I can't really test the OOM scenario

You don't need to test it.
There's no OOM scenario as long as the c++ does the same than the lua version.
Just massive speed gains :).

>> it actually looks like it's actually the sheer amount of k,v pairs that just balloons to stupid amounts

> Actually... large string values could explain some of the overhead (e.g., the most common one being description).

Yup, too much to store the whole DOM on memory. The vm just dies.

**Comment by pazos** [comment]
>> I'd rather this make it in, actually.

> Although, perhaps with slighty different menu labels/config keys. I'm thinking something along the lines of:

> Fast (RapidJSON via load_calibre)
> Safe (LunaJSON)
> Legacy (RapidJSON via load)
> (And, for now, until we have more data from insane libraries, keep the size check, but perhaps default to fast in order to exercise the codepath and find potential issues with it).

The option to skip UI and go reckless on the fast rapidjson codepath is out of the equation?.

**Comment by NiLuJe** [comment]
> The option to skip UI and go reckless on the fast rapidjson codepath is out of the equation?

I mean, eventually, maybe? Right now, I'd keep it in just to be safe ;p.

**Review by pazos — COMMENTED** [review]
Good!

Feel free to bump base and merge when ready :)

**Comment by benoit-pierre** [comment]
The `rapidjson.load_calibre` variant crashes on my test data, here is the first entry:
```json
    {
        "series_index": null,
        "size": 2520019,
        "series": null,
        "last_modified": "2023-10-12T08:41:34+00:00",
        "authors": [
            "Stephen King"
        ],
        "tags": {},
        "lpath": "Shining, The - Stephen King (2279).epub",
        "title": "The Shining"
    },
```

I don't know how and/or why it ended up with `"tags": {},` instead of `"tags": [],`, but this cause the crash.

**Comment by pazos** [comment]
> I don't know how and/or why it ended up with "tags": {}, instead of "tags": [],, but this cause the crash.

Do you use the wireless client? Because it rewrites the json file based on current json parser. So any issue with the parser ends up in the original json file.

If that's true, then the metadata itself isn't valid anymore. During my tests with the lua parser I went the following route to avoid these kind of issues.

1. Delete json files (and books)
2. Push books and json files using calibre's connect to folder
3. Verify the metadata can be parsed and searches work ok with any field
4. Disconnect from folder, start the wireless client
5. Verify the client comm with the server is ok
6. Send a few books wirelessly
7. Verify the metadata is still correct.

**Comment by benoit-pierre** [comment]
I had previously connected to calibre with the wireless client, but I don't know if the bad data is from before or after updating to test the new code.

**Comment by benoit-pierre** [comment]
Anyway, it should not crash.

**Comment by NiLuJe** [comment]
> but may not take too kindly to malformed input files

Famous last words ;p.

----

Yeah, that probably unbalances the push/pop on the Lua stack, trashing the stack.

I'll see if I can come up with something not too awful to not horribly implode ;).

**Comment by NiLuJe** [comment]
@pazos: Do you remember if the frontend code can deal with potential array fields being `nil`?

(i.e., should I push a `nil` or a `{}` for these broken "arrays"?).

**Comment by NiLuJe** [comment]
Alternatively, how does it deal with a *missing* required field? (as that's my other potential approach ;p).

**Comment by pazos** [comment]
@NiLuJe: There's https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/metadata.lua#L41-L53, which should turn nils to whatever value calibre expects.

The only values that are required are the strings, which should never be nill (calibre will fill bogus values if it can figure out the proper ones, i.e: last_modified: None)

**Comment by NiLuJe** [comment]
> I don't know how and/or why it ended up with `"tags": {},` instead of `"tags": [],`

That's because when dumping back to json, for tables that have lost their metatable array/object flag, by default, it assumes emtpoy tables are *objects*, not tables (c.f., `empty_table_as_array` option).

And the aforementioned `slim` method, when confronted with a missing field, sets arrays to an untagged empty table ;).

On the same subject, I'd removed those tags from the load_calibre variant, on the assumption that these could never end up beign dumped back to json, but that might not have been a great idea ;).
